### DISCUSSION TITLE
NRPN, beyond pitchbend and PBPotentiometer

### DISCUSSION THREAD

**Discussion by FotisDovas** [tttapa/Control-Surface#639] [Q&A]
Hello dear people!
Hello tttapa!

I was wondering, if there is a way to implement the control surface library, so as to combine the midiCc 0-31 (msb) with midiCc 32-63 (lsb), so as to create more nrpnPotentiometers.

Seems like 16 channels * 16 cables * numberOfPots will not be sufficient, as i ideally would like to have many, many banks..  : /
xexe

So, would you say that there is a way to use the already existing support+ code of PBPotentiometer, 
but with the addition of applying it to cc0 - cc63 combos?

Thank you!

**Comment by tttapa** [comment]
> so as to combine the midiCc 0-31 (msb) with midiCc 32-63 (lsb)

This doesn't sound like NRPN to me, but it is supported by the library: https://github.com/tttapa/Control-Surface/issues/202#issuecomment-634034911

> Seems like 16 channels * 16 cables * numberOfPots will not be sufficient

Are you sure? That's 8192 14-bit controls!

Sending NRPNs is supported as well, but you'll have to send the MIDI messages yourself, e.g. using `Control_Surface.sendControlChange(...)`.

**Reply by FotisDovas** [reply]
hmm, i guess i missed some things on midi theory.

The machine i am building will have something like 6*12 pots, and about 12 banks. 
I thought i can only send 14bit values through pitchbend, which is, in my head, constrained to 1 per midi_channel. 

May the solution you are suggesting work with the turning of one potentiometer? which will combine the msb-lsbs ? Or should I use 2 potentiometers?
I am saying that now because i did not make this clear at the former post.. and will be digging thoroughly

If so, i will pay all my attention to your code and find the way again  : )

Did I mention you are one of the best? Thank you

**Reply by FotisDovas** [reply]
omg, cc14, you are a god

(continuousCc14sender)

**Reply by FotisDovas** [reply]
Hello Again!

You think there is a way to get the ContinuousCCSender bitRate from 10 to 14, without the 1/2000 noise it has now ?
Just so that i can map the controls to the DAW with the simple "link variable to last tweaked midiControl"
instead of manualy.

Thanks

**Reply by Veroledez** [reply]
Hola, que placa estas usando?
el ruido se debe a la resolucion, ya que esta pasando de 10 bits o 12 bits segun la placa, a 14 bits, por lo que faltan valores, distinto a cuando pasas a 7 bits que disminuyes la resolucion  y da margen de ruido. 
Para estabilizar un poco puedes poner un condensador de 0.1uf

**Reply by FotisDovas** [reply]
Very well said my friend.
I prefered a Teensy 3.5 board, which had a resolute ADC, at 13bits.
What i thought untill now, was that the bit resolution was increased to 14 through the use of code in the CCSender.

Would you say that the capacitor (great idea, i am new to electronics), would deliver stability, even if the bit resolution increases with the use of code?  

+ right now, my wiring is less than adequate

**Reply by Veroledez** [reply]
Para 14 bits te recomiendo usar un adc externo de 16 bits,  yo uso el condensador en esp32 y estabiliza bastante.

Puedes probar con el ejemplo de FilteredAnalog, hay una parte que usa varias lecturas para mayor estabilidad, yo he logrado buena estabilidad en 12 bits usando 12 lecturas. 

Hay que tener en cuenta la tolerancia del potenciometro, unos dan mucho ruido debido a que la tolerancia es +-5, mientras que hay de +-2 que dan menos ruido. Tambien influye el tamaño de este, ya que en uno de 60mm da mas ruido que uno de 100mm, por esos las superficies de control grandes como presonus, qcon usan 100mm es sus faders

**Reply by FotisDovas** [reply]
very good, my friend! I'll try my best!
